packages/nextjs/src/config/manifest/buildManifest.ts

```diff
@@ -0,0 +1,196 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+export type RouteInfo = {
+  path: string;
+  dynamic: boolean;
+  pattern?: string;
+  paramNames?: string[];
+};
+
+export type RouteManifest = {
+  dynamic: RouteInfo[];
+  static: RouteInfo[];
+};
+
+export type CreateRouteManifestOptions = {
+  // For starters we only support app router
+  appDirPath?: string;
+};
+
+let manifestCache: RouteManifest | null = null;
+let lastAppDirPath: string | null = null;
+
+function isPageFile(filename: string): boolean {
+  return filename === 'page.tsx' || filename === 'page.jsx' || filename === 'page.ts' || filename === 'page.js';
+}
+
+function isRouteGroup(name: string): boolean {
+  return name.startsWith('(') && name.endsWith(')');
+}
+
+function getDynamicRouteSegment(name: string): string {
+  if (name.startsWith('[[...') && name.endsWith(']]')) {
+    // Optional catchall: [[...param]]
+    const paramName = name.slice(5, -2); // Remove [[... and ]]
+    return `:${paramName}*?`; // Mark with ? as optional
+  } else if (name.startsWith('[...') && name.endsWith(']')) {
+    // Required catchall: [...param]
+    const paramName = name.slice(4, -1); // Remove [... and ]
+    return `:${paramName}*`;
+  } else {
+    // Regular dynamic: [param]
+    return `:${name.slice(1, -1)}`;
+  }
```

Comment:
Is there any reason to have a `else` here, since there is no more code to the function after that? 🙂

```suggestion
  } 
 // Regular dynamic: [param]
 return `:${name.slice(1, -1)}`;
```